File: src/backend/ponto/src/model/ponto_parametro_model.py
from src import db


class PontoParametroModel(db.Model):
    __tablename__ = 'ponto_parametro'

    id = db.Column(db.Integer, primary_key=True)
    mes_ano = db.Column(db.String(7))
    dia_inicial_apuracao = db.Column(db.Integer)
    hora_noturna_inicio = db.Column(db.String(8))
    hora_noturna_fim = db.Column(db.String(8))
    periodo_minimo_interjornada = db.Column(db.String(8))
    percentual_he_diurna = db.Column(db.Float)
    percentual_he_noturna = db.Column(db.Float)
    duracao_hora_noturna = db.Column(db.String(8))
    tratamento_hora_mais = db.Column(db.String(1))
    tratamento_hora_menos = db.Column(db.String(1))


    def mapping(self, data):
        self.id = data.get('id')
        self.mes_ano = data.get('mesAno')
        self.dia_inicial_apuracao = data.get('diaInicialApuracao')
        self.hora_noturna_inicio = data.get('horaNoturnaInicio')
        self.hora_noturna_fim = data.get('horaNoturnaFim')
        self.periodo_minimo_interjornada = data.get('periodoMinimoInterjornada')
        self.percentual_he_diurna = data.get('percentualHeDiurna')
        self.percentual_he_noturna = data.get('percentualHeNoturna')
        self.duracao_hora_noturna = data.get('duracaoHoraNoturna')
        self.tratamento_hora_mais = data.get('tratamentoHoraMais')
        self.tratamento_hora_menos = data.get('tratamentoHoraMenos')

    def serialize(self):
        return {
            'id': self.id,
            'mesAno': self.mes_ano,
            'diaInicialApuracao': self.dia_inicial_apuracao,
            'horaNoturnaInicio': self.hora_noturna_inicio,
            'horaNoturnaFim': self.hora_noturna_fim,
            'periodoMinimoInterjornada': self.periodo_minimo_interjornada,
            'percentualHeDiurna': self.percentual_he_diurna,
            'percentualHeNoturna': self.percentual_he_noturna,
            'duracaoHoraNoturna': self.duracao_hora_noturna,
            'tratamentoHoraMais': self.tratamento_hora_mais,
            'tratamentoHoraMenos': self.tratamento_hora_menos,
        }

File: frontend/ponto/lib/app/data/model/ponto_parametro_model.dart
import 'dart:convert';
import 'package:pluto_grid/pluto_grid.dart';

import 'package:ponto/app/infra/infra_imports.dart';
import 'package:ponto/app/data/domain/domain_imports.dart';

class PontoParametroModel {
	int? id;
	String? mesAno;
	int? diaInicialApuracao;
	String? horaNoturnaInicio;
	String? horaNoturnaFim;
	String? periodoMinimoInterjornada;
	double? percentualHeDiurna;
	double? percentualHeNoturna;
	String? duracaoHoraNoturna;
	String? tratamentoHoraMais;
	String? tratamentoHoraMenos;

	PontoParametroModel({
		this.id,
		this.mesAno,
		this.diaInicialApuracao,
		this.horaNoturnaInicio,
		this.horaNoturnaFim,
		this.periodoMinimoInterjornada,
		this.percentualHeDiurna,
		this.percentualHeNoturna,
		this.duracaoHoraNoturna,
		this.tratamentoHoraMais,
		this.tratamentoHoraMenos,
	});

	static List<String> dbColumns = <String>[
		'id',
		'mes_ano',
		'dia_inicial_apuracao',
		'hora_noturna_inicio',
		'hora_noturna_fim',
		'periodo_minimo_interjornada',
		'percentual_he_diurna',
		'percentual_he_noturna',
		'duracao_hora_noturna',
		'tratamento_hora_mais',
		'tratamento_hora_menos',
	];
	
	static List<String> aliasColumns = <String>[
		'Id',
		'Mes Ano',
		'Dia Inicial Apuracao',
		'Hora Noturna Inicio',
		'Hora Noturna Fim',
		'Periodo Minimo Interjornada',
		'Percentual He Diurna',
		'Percentual He Noturna',
		'Duracao Hora Noturna',
		'Tratamento Hora Mais',
		'Tratamento Hora Menos',
	];

	PontoParametroModel.fromJson(Map<String, dynamic> jsonData) {
		id = jsonData['id'];
		mesAno = jsonData['mesAno'];
		diaInicialApuracao = jsonData['diaInicialApuracao'];
		horaNoturnaInicio = jsonData['horaNoturnaInicio'];
		horaNoturnaFim = jsonData['horaNoturnaFim'];
		periodoMinimoInterjornada = jsonData['periodoMinimoInterjornada'];
		percentualHeDiurna = jsonData['percentualHeDiurna']?.toDouble();
		percentualHeNoturna = jsonData['percentualHeNoturna']?.toDouble();
		duracaoHoraNoturna = jsonData['duracaoHoraNoturna'];
		tratamentoHoraMais = PontoParametroDomain.getTratamentoHoraMais(jsonData['tratamentoHoraMais']);
		tratamentoHoraMenos = PontoParametroDomain.getTratamentoHoraMenos(jsonData['tratamentoHoraMenos']);
	}

	Map<String, dynamic> get toJson {
		Map<String, dynamic> jsonData = <String, dynamic>{};

		jsonData['id'] = id != 0 ? id : null;
		jsonData['mesAno'] = Util.removeMask(mesAno);
		jsonData['diaInicialApuracao'] = diaInicialApuracao;
		jsonData['horaNoturnaInicio'] = Util.removeMask(horaNoturnaInicio);
		jsonData['horaNoturnaFim'] = Util.removeMask(horaNoturnaFim);
		jsonData['periodoMinimoInterjornada'] = Util.removeMask(periodoMinimoInterjornada);
		jsonData['percentualHeDiurna'] = percentualHeDiurna;
		jsonData['percentualHeNoturna'] = percentualHeNoturna;
		jsonData['duracaoHoraNoturna'] = Util.removeMask(duracaoHoraNoturna);
		jsonData['tratamentoHoraMais'] = PontoParametroDomain.setTratamentoHoraMais(tratamentoHoraMais);
		jsonData['tratamentoHoraMenos'] = PontoParametroDomain.setTratamentoHoraMenos(tratamentoHoraMenos);
	
		return jsonData;
	}
	
	String objectEncodeJson() {
		final jsonData = toJson;
		return json.encode(jsonData);
	}

	plutoRowToObject(PlutoRow plutoRow) {
		id = plutoRow.cells['id']?.value;
		mesAno = plutoRow.cells['mesAno']?.value;
		diaInicialApuracao = plutoRow.cells['diaInicialApuracao']?.value;
		horaNoturnaInicio = plutoRow.cells['horaNoturnaInicio']?.value;
		horaNoturnaFim = plutoRow.cells['horaNoturnaFim']?.value;
		periodoMinimoInterjornada = plutoRow.cells['periodoMinimoInterjornada']?.value;
		percentualHeDiurna = plutoRow.cells['percentualHeDiurna']?.value?.toDouble();
		percentualHeNoturna = plutoRow.cells['percentualHeNoturna']?.value?.toDouble();
		duracaoHoraNoturna = plutoRow.cells['duracaoHoraNoturna']?.value;
		tratamentoHoraMais = plutoRow.cells['tratamentoHoraMais']?.value != '' ? plutoRow.cells['tratamentoHoraMais']?.value : 'Extra';
		tratamentoHoraMenos = plutoRow.cells['tratamentoHoraMenos']?.value != '' ? plutoRow.cells['tratamentoHoraMenos']?.value : 'Falta';
	}	

	PontoParametroModel clone() {
		return PontoParametroModel(
			id: id,
			mesAno: mesAno,
			diaInicialApuracao: diaInicialApuracao,
			horaNoturnaInicio: horaNoturnaInicio,
			horaNoturnaFim: horaNoturnaFim,
			periodoMinimoInterjornada: periodoMinimoInterjornada,
			percentualHeDiurna: percentualHeDiurna,
			percentualHeNoturna: percentualHeNoturna,
			duracaoHoraNoturna: duracaoHoraNoturna,
			tratamentoHoraMais: tratamentoHoraMais,
			tratamentoHoraMenos: tratamentoHoraMenos,
		);			
	}

	
}